File: stack.js
// Stack implemented using linked list as internal data structure
require './singlyLinkedList.js';

class Stack {
  constructor() {
    this.list = new LinkedList;
    this.least = null;
    this.most = null;
    this.minList = new LinkedList;
    this.maxList = new LinkedList;
    this.size = this.list.length;
  }
  push(data) {
    this.list.push(data);
    this.least === null || this.minList.head === null || data < this.least ? (this.minList.commit(data), this.least = data) : undefined;
    this.most === null || this.maxList.head === null || data > this.most ? (this.maxList.commit(data), this.most = data) : undefined;
    this.size++;
    return this;
  }
  pop() {
    const popped = this.list.pop();
    popped === this.minList.head.data ? (this.minList.reset(), this.least = this.minList.head === null ? null : this.minList.head.data) : undefined;
    popped === this.maxList.head.data ? (this.maxList.reset(), this.most = this.maxList.head === null ? null : this.maxList.head.data) : undefined;
    this.size--;
    return popped;
  }
  min() {
    return this.least;
  }
  max() {
    return this.most;
  }
}

const stack = new Stack,
  input = [3, 5, -1, -1, 2, 7, 11, -1, -1];

function do_stuff(arr) {
  let [temp, output] = [
    [], ""
  ];

  arr.forEach(v => v === -1 ? output += " " + temp.pop() : temp.push(v));

  return output;
}

File: stack.rb
# Stack implemented using linked list as internal data structure
require_relative 'singly_linked_list'

class Stack
  attr_reader :size
  def initialize
    @list = LinkedList.new
    @least = nil
    @most = nil
    @min_list = LinkedList.new
    @max_list = LinkedList.new
    @size = @list.length
  end

  def push(data)
    @list.push(data)
    # Reassigning minimum
    if @least.nil? || @min_list.head.nil? || data < @least 
      @min_list.commit(data)
      @least = data
    else
      nil
    end
    # Reassigning maximum
    if @most.nil? || @max_list.head.nil? || data > @most 
      @max_list.commit(data)
      @most = data
    else
      nil
    end
    @size += 1
    self
  end

  def pop
    popped = @list.pop
    # Reassign minimum when popping
    if popped == @min_list.head.data
      @min_list.reset
      @least = @min_list.head.nil? ? nil : @min_list.head.data
    else
      nil
    end
    # Reassigning maximum when popping
    if popped == @max_list.head.data
      @max_list.reset
      @most = @max_list.head.nil? ? nil : @max_list.head.data
    else
      nil
    end
    @size -= 1
    popped
  end

  def min
    @least
  end

  def max
    @most
  end

  def peek
    self.is_empty? ? "No data in the stack" : @list.head.data
  end

  def is_empty?
    @list.length == 0
  end

end

stack = Stack.new
input = [3,5,-1,-1,2,7,11,-1,-1]

def do_stuff(arr)
  tmp, output = Stack.new, ""
  
  arr.each { |el| el == -1 ?  output += tmp.pop.to_s + " ": tmp.push(el)}

  output
end

# puts do_stuff(input)